### PR TITLE
New, smarter fast_fill function

### PR DESCRIPTION
**Pull request by Alchav**
Currently, if a location has an item rule that prevents a non-advancement, non-never-exclude item from being allowed, this only causes failed generation instead of intelligently avoiding placing these items in these locations.

The local_items and non_local_items options for item links works by placing a lot of item rules, and so using these options on any non-advancement items is likely to cause failed generation.

In addition to fixing that issue, this will allow for other item rules affecting junk items.

### CODE REVIEW

**Comment by Berserker66** [comment]
Can you provide a yaml and seed to reproduce this issue and test the proposed fix against?

**Comment by Alchav** [comment]
-- seed 1

`name: A
game: A Link to the Past
A Link to the Past:
  item_links: [{name: Blue Rupee Link, item_pool: ["Rupees (5)"], local_items: ["Rupees (5)"], replacement_item: null}]
---
name: B
game: A Link to the Past
A Link to the Past:
  item_links: [{name: Blue Rupee Link, item_pool: ["Rupees (5)"], local_items: ["Rupees (5)"], replacement_item: null}]
---
name: C
game: A Link to the Past
A Link to the Past:
 {}`

**Comment by Alchav** [comment]
> it might be possible to refactor has_item_rule into a property of `self.item_rule is not self.__class__.item_rule` or similar, would need to check.

Appears to work.

**Comment by Berserker66** [comment]
This looks to be a bug with @CaitSith2 's change to ItemLinks now allowing local items, but they're not added to the local item fill pools, resulting in them falling through to fast fill.

**Comment by Alchav** [comment]
The item link local/non-local items could be handled much more efficiently, but I think this change should be made regardless so that locations can forbid non-advancement items in the future.

**Comment by Berserker66** [comment]
That's supposed to go through excluded locations. Full logical fill is fairly slow, which is why specific fill stages exist that speed up their special case. I know the easy solution is to just fill everything with logic and nothing would break ever.

**Comment by Berserker66** [comment]
Alternative way to go about it could be to get all locations with item rules and make a specialized fill for them, and seeing what other "old" fill stages might be clumped into it without additional fill-time overhead.

**Comment by black-sliver** [comment]
I think the proper approach for the current system/code would be to detect (or define) which items may be restricted by item_rules and fill them separately or as part of the slow fill. This is a bit annoying for dynamically generated junk items with local and nonlocal, but filling all items restrictive does not work very well. I did that in the early SoE days for testing.

**Comment by Alchav** [comment]
> Alternative way to go about it could be to get all locations with item rules and make a specialized fill for them, and seeing what other "old" fill stages might be clumped into it without additional fill-time overhead.

I think I can do this.

File: Fill.py [hunk 1]
```diff
@@ -136,33 +136,93 @@ def fill_restrictive(world: MultiWorld, base_state: CollectionState, locations:
     itempool.extend(unplaced_items)
 
 
+def fast_fill(world: MultiWorld, locations: typing.List[Location],
+                     itempool: typing.List[Item], single_player_placement: bool = False) -> None:
+    unplaced_items: typing.List[Item] = []
+    placements: typing.List[Location] = []
+    swap = 0
+    swapped_items: typing.Counter[typing.Tuple[int, str]] = Counter()
+    while locations and itempool:
```
Comment (Berserker66) [inline Fill.py:145]: the fill then swap code looks very similar to the other one; maybe that can be turned into a common function?

**Comment by Berserker66** [comment]
Any benchmarks that compare this to the current algorithm for a few games? Especially long gamess like HK and OoT would be good.

**Comment by Alchav** [comment]
> Any benchmarks that compare this to the current algorithm for a few games? Especially long gamess like HK and OoT would be good.

On my laptop with spoilers generation off, using `--seed 2` and this yaml:
```
name: HK1
game: Hollow Knight
Hollow Knight:
  RandomizePaleOre: True
  RandomizeMaskShards: True
  RandomizeVesselFragments: True
  RandomizeGeoRocks: True
  RandomizeRancidEggs: True
  local_items: ["Geo_Rock-Default"]
  non_local_items: ["Rancid_Egg"]
---
name: HK2
game: Hollow Knight
Hollow Knight:
  RandomizePaleOre: True
  RandomizeMaskShards: True
  RandomizeVesselFragments: True
  RandomizeGeoRocks: True
  RandomizeRancidEggs: True
  local_items: ["Geo_Rock-Default"]
  non_local_items: ["Rancid_Egg"]
---
name: OoT
game: Ocarina of Time
Ocarina of Time:
  local_items: ["Rupees (20)"]
  non_local_items: ["Rupees (5)"]
```
On Main:
`Done. Enjoy. Total Time: 207.3636798`

On this branch:
`Done. Enjoy. Total Time: 160.79275760000002`

File: Fill.py (hunk 1 above)
Comment (Alchav) [reply Fill.py:145]: I'm not seeing what could easily be done to that end, every section has differences

File: Fill.py [hunk 2]
```diff
@@ -136,33 +136,92 @@ def fill_restrictive(world: MultiWorld, base_state: CollectionState, locations:
     itempool.extend(unplaced_items)
 
 
+def fast_fill(world: MultiWorld, locations: typing.List[Location],
+                     itempool: typing.List[Item], single_player_placement: bool = False) -> None:
+    unplaced_items: typing.List[Item] = []
+    placements: typing.List[Location] = []
+    swap = 0
```
Comment (Berserker66) [inline Fill.py:143]: swap gets set to 0 and then incremented later, but nothing is ever done with this value.

**Comment by Berserker66** [comment]
My attempt to improve on this:
https://github.com/Alchav/Archipelago/pull/6

**Review by Berserker66 — APPROVED** [review]

**Pull request merged**